### PR TITLE
Enhance: upgrade dubbogo URL

### PR DESCRIPTION
**Pull request by georgehao**
**What this PR does**:
Enhance dubbogo URL

**Which issue(s) this PR fixes**:

I have read some discuss [How to copy a struct which contains a mutex](https://groups.google.com/g/golang-nuts/c/imxjBLNJ9OY). The discuss tell us can't copy a struct which contains `sync` type or variable. 

And the struct contains `sync` type can see anywhere of source code of k8s, but k8s use the struct with pointer. eg: [Cache](https://github.com/kubernetes/kubernetes/blob/d20e3246bade17564860655e7fa57ead922d7307/pkg/controller/endpointslicemirroring/metrics/cache.go#L37)

I have two way to fix the URL.

#### **way1** : move the `paramsLock  sync.RWMutex` to `paramsLock   *sync.RWMutex`

like this:
```
type URL struct {
        // ....
        Path     string
        paramsLock   *sync.RWMutex
}
```

it has some problem:
1. when new URL, we must init the `paramsLock`, otherwise may call the nil pointer panic.
2. some struct may share the same `paramsLock`, will influence each other. 

so it's a bad solution.

#### **way2**: make `common.URL` noCopy. must use the `*common.URL` instead of the `common.URL`. 

this way is the general way. so i select this

I also add `noCopy` struct to `common.URL`, it's will warning us not copy the URL struct. 

#### other
my solution is  similar to the PR #773, but #773 missing some files not modify.
 
I have add the feature of #823 do.

### CODE REVIEW

**Review by zouyx — COMMENTED** [review]
Thanks @georgehao , make `Checks`normal pls.
BTW can i close #773 ?

**Comment by georgehao** [comment]
> Thanks @georgehao , make `Checks`normal pls.
> BTW can i close #773 ?

yes, you can close 773.

File: common/proxy/proxy_factory/default.go [hunk 1]
```diff
@@ -90,7 +90,7 @@ func (pi *ProxyInvoker) Invoke(ctx context.Context, invocation protocol.Invocati
 
 	url := pi.GetUrl()
 	//get providerUrl. The origin url may be is registry URL.
-	url = *getProviderURL(&url)
+	url = getProviderURL(url)
```
Comment (cvictory) [inline common/proxy/proxy_factory/default.go:93]: 两行可以合并吧

File: go.mod [hunk 2]
```diff
@@ -51,6 +51,7 @@ require (
 
 replace (
 	github.com/envoyproxy/go-control-plane => github.com/envoyproxy/go-control-plane v0.8.0
+	github.com/shirou/gopsutil => github.com/shirou/gopsutil v0.0.0-20181107111621-48177ef5f880
```
Comment (cvictory) [inline go.mod:54]: 为什么指定版本？ 而且是通过replace方式？

File: config/config_loader_test.go [hunk 3]
```diff
@@ -418,23 +417,23 @@ func (mts *mockMetadataService) addURL(targetMap *sync.Map, url *common.URL) boo
...
-func (m *mockMetadataService) getAllService(services *sync.Map) []common.URL {
+func (m *mockMetadataService) getAllService(services *sync.Map) []*common.URL {
 	// using skip list to dedup and sorting
-	res := make([]common.URL, 0)
+	var res []*common.URL
 	services.Range(func(key, value interface{}) bool {
 		urls := value.(*skip.SkipList)
 		for i := uint64(0); i < urls.Len(); i++ {
-			url := common.URL(urls.ByPosition(i).(Comparator))
+			url := urls.ByPosition(i).(*common.URL)
```
Comment (cvictory) [inline config/config_loader_test.go:436]: 转换确认下，没有问题。

File: go.mod (hunk 2 above)
Comment (georgehao) [reply go.mod:54]: @zouyx 贤哥搞的, 贤哥看下吧

File: config/config_loader_test.go (hunk 3 above)
Comment (georgehao) [reply config/config_loader_test.go:436]: 嗯, 我用URL实现了Comparator, 把config_loader里面实现的Comparator给去掉了

File: common/proxy/proxy_factory/default.go (hunk 1 above)
Comment (georgehao) [reply common/proxy/proxy_factory/default.go:93]: OK

**Comment by fangyincheng** [comment]
> > Thanks @georgehao , make `Checks`normal pls.
> > BTW can i close #773 ?
> 
> yes, you can close 773.

what about 823?

**Review by AlexStocks — APPROVED** [review]

File: go.mod (hunk 2 above)
Comment (AlexStocks) [reply go.mod:54]: https://github.com/alibaba/sentinel-golang/issues/326

**Review by hxmhlt — APPROVED** [review]

**Review by fangyincheng — APPROVED** [review]

**Pull request merged**